Faster guid<->string

christopherlam opened the pull request:
specialises 32-byte string<->guid. benchmark commit shown but not to be merged in. x4 to x16 times faster.

before:
```
Executed 100249 tests. All tests passed.
guid_to_string: 11818 ms
guid_to_string_buff: 12517 ms
string_to_guid: 19453 ms
```

after:
```
Executed 100249 tests. All tests passed.
guid_to_string: 750 ms
guid_to_string_buff: 649 ms
string_to_guid: 2999 ms
```

christopherlam committed:
[guid.cpp] fix guid==nullptr comparison retvals

- if guid_1 guid_2 are both null they should return int(0) or bool(true)

jralls commented on libgnucash/engine/guid.cpp:
```diff
@@ -164,30 +164,58 @@ gchar *
+char_to_num (unsigned char c) noexcept
+{
+    unsigned int digit = c - '0';
+    unsigned int alpha = (c | 0x20) - 'a';
```
Add handling `A-F`:
```
unsigned int alpha = c >= 'A' ? c - 'A' : (c | 0x20) - 'a';
```

christopherlam replied:
> Add handling `A-F`:
> 
> ```
> unsigned int alpha = c >= 'A' ? c - 'A' : (c | 0x20) - 'a';
> ```

The `c|0x20` does it?

jralls replied:
Ah, so it does. Cool.

jralls commented on libgnucash/engine/guid.cpp:
```diff
@@ -164,30 +164,58 @@ gchar *
 {
     if (!guid || !str || !*str) return false;
 
+    if (std::memchr (str, '\0', 33) == str + 32 && fast_string_to_guid (str, guid))
```
`strlen(str) == 32` is clearer and does the same thing as `memchr`.

jralls reviewed (changes requested):
This is going backwards for C++ conversion so optimize the C++ functions instead and leave the C functions to wrap them. The mass users of these functions, the backends and CSV import/export, are already converted to C++ so they could be calling the `gnc::GUID` functions directly, which would save some of the allocating and copying.

jralls commented on libgnucash/engine/test/test-guid.cpp:
```diff
@@ -93,6 +96,73 @@ run_test (void)
+    auto duration = duration_cast<milliseconds>(end - start);
+    std::cout << "string_to_guid: " << duration.count() << " ms\n";
+}
+
```
```
-- a/libgnucash/engine/test/test-guid.cpp
+++ b/libgnucash/engine/test/test-guid.cpp
@@ -28,6 +28,7 @@
 #include <guid.hpp>
 #include <glib.h>

+#include <array>
 #include <vector>
 #include <memory>
 #include <chrono>
@@ -145,17 +146,17 @@ static void benchmark_guid_to_string_buff() {

 static void benchmark_string_to_guid() {
     std::vector<GncGUID> guids(1000);
-    std::vector<char[GUID_ENCODING_LENGTH + 1]> buffers(1000);
+    std::vector<std::array<char, GUID_ENCODING_LENGTH + 1>> buffers(1000);

     for (size_t  i = 0; i < guids.size(); ++i) {
         guids[i] = random_guid();
-        guid_to_string_buff(&guids[i], buffers[i]);
+        guid_to_string_buff(&guids[i], buffers[i].data());
     }

     GncGUID g;
     auto start = high_resolution_clock::now();
     for (size_t i = 0; i < NUM_ITER; ++i) {
-        char* str = buffers[i % buffers.size()];
+        char* str = buffers[i % buffers.size()].data();
         string_to_guid(str, &g);
     }
     auto end = high_resolution_clock::now();
```
Gets it to pass on macOS.

jralls commented:
Looks good.

christopherlam committed:
[guid.cpp] Specialise guid<->32-byte string

- string->guid has branch-lite unrollable loop, fallback to slow
  boost parser
- guid->string uses lookup table
- avoids creating several temporary heap-allocated std::strings

christopherlam commented:
benchmarking commit removed, pasted here:

```diff
diff --git a/libgnucash/engine/test/test-guid.cpp b/libgnucash/engine/test/test-guid.cpp
index 78a4e5c4f6..0c0f88615d 100644
--- a/libgnucash/engine/test/test-guid.cpp
+++ b/libgnucash/engine/test/test-guid.cpp
@@ -29,7 +29,11 @@
 #include <glib.h>
 
 #include <vector>
+#include <array>
 #include <memory>
+#include <chrono>
+#include <iostream>
+#include <cassert>
 
 #include <config.h>
 #include <ctype.h>
@@ -93,6 +97,73 @@ run_test (void)
     qof_session_destroy(sess);
 }
 
+
+using namespace std::chrono;
+
+const int NUM_ITER = 10'000'000;  // 10M iterations for microbenchmark
+
+// Helper to generate test GUID
+static GncGUID random_guid()
+{
+    GncGUID g;
+    for (size_t i = 0; i < GUID_DATA_SIZE; ++i)
+        g.reserved[i] = rand() & 0xFF;
+    return g;
+}
+
+static void benchmark_guid_to_string() {
+    std::vector<GncGUID> guids(1000);
+    for (auto &g : guids) g = random_guid();
+
+    auto start = high_resolution_clock::now();
+    for (size_t i = 0; i < NUM_ITER; ++i) {
+        GncGUID &g = guids[i % guids.size()];
+        char *str = guid_to_string(&g);
+        assert(str != nullptr);  // sanity check
+        g_free(str);
+    }
+    auto end = high_resolution_clock::now();
+    auto duration = duration_cast<milliseconds>(end - start);
+    std::cout << "guid_to_string: " << duration.count() << " ms\n";
+}
+
+static void benchmark_guid_to_string_buff() {
+    std::vector<GncGUID> guids(1000);
+    for (auto &g : guids) g = random_guid();
+
+    char buffer[GUID_ENCODING_LENGTH + 1];
+
+    auto start = high_resolution_clock::now();
+    for (size_t  i = 0; i < NUM_ITER; ++i) {
+        GncGUID &g = guids[i % guids.size()];
+        guid_to_string_buff(&g, buffer);
+        assert(strlen(buffer) == GUID_ENCODING_LENGTH);
+    }
+    auto end = high_resolution_clock::now();
+    auto duration = duration_cast<milliseconds>(end - start);
+    std::cout << "guid_to_string_buff: " << duration.count() << " ms\n";
+}
+
+static void benchmark_string_to_guid() {
+    std::vector<GncGUID> guids(1000);
+    std::vector<std::array<char, GUID_ENCODING_LENGTH + 1>> buffers(1000);
+
+    for (size_t  i = 0; i < guids.size(); ++i) {
+        guids[i] = random_guid();
+        guid_to_string_buff(&guids[i], buffers[i].data());
+    }
+
+    GncGUID g;
+    auto start = high_resolution_clock::now();
+    for (size_t i = 0; i < NUM_ITER; ++i) {
+        char* str = buffers[i % buffers.size()].data();
+        string_to_guid(str, &g);
+    }
+    auto end = high_resolution_clock::now();
+    auto duration = duration_cast<milliseconds>(end - start);
+    std::cout << "string_to_guid: " << duration.count() << " ms\n";
+}
+
 int
 main (int argc, char **argv)
 {
@@ -104,5 +175,10 @@ main (int argc, char **argv)
         print_test_results();
     }
     qof_close();
+
+    benchmark_guid_to_string();
+    benchmark_guid_to_string_buff();
+    benchmark_string_to_guid();
+
     return get_rv();
 }
```
